jsonpath: allow case insensitivity for keyword ident

ZhouXing19 opened the pull request:
Epic: None
Fix: https://github.com/cockroachdb/cockroach/issues/144255

Previously, if the keyword is not all capital letter, it will cause syntax error. For example, `strIcT $.STRiCt == null` will failed the parsing with `at or near "strIcT": syntax error`. This is because the lexer sees `strIcT $.STRiCt == null` as `IDENT $.IDENT == NULL`, which is a rule that is not found in `jsonpath.y`. 

Note that the lexer recognizes `strIcT` and `STRiCt` as `IDENT` by [the id](https://github.com/cockroachdb/cockroach/blob/ac3b2b749ca61a781566557faf29daa08ae2f9b5/pkg/util/jsonpath/parser/lexer.go#L69-L79), not by the string. So this commit is to set the ID for all string whose lowercased form matches a keyword. More specifically, we now do `lval.SetID(lexbase.GetKeywordID(strings.ToLower(lval.Str())))` in the scanning logic.

So with the same example, `strIcT $.STRiCt == null` is now recognized as `STRICT $.STRICT == NULL` by the parser, for which a rule exists in the yacc and it will pass the **Lexical Analysis**. Then, at the **Syntactic Analysis** step, the keyword `STRICT` will be normalized, while the property accessing part, `$.STRiCt`, will remain the same, as the original str `STRiCt` is still reserved.

Release note (bug fix): Support any casing for keyword identifiers in a jsonpath query. Note that for 3 special
identifiers "TRUE", "FALSE", "NULL", crdb parses them with case insensitivity, but PG parses them case sensitively, e.g. `select '$.active == TrUe'::jsonpath;` will work in crdb but not in pg.

michae2 reviewed (commented):
Does this change the result of a query like `SELECT jsonb_path_query('{"a": 1, "A": 2}', '$.A');`?

> Reviewed 2 of 3 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/149251#-:-OU1GAEO2Wvh6og8A1Vd:bfbigjx)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @mgartner)



<!-- Sent from Reviewable.io -->

mgartner reviewed (commented):
Ya, I don't think this change matches Postgres behavior. JSONPATH is case sensitive:

```
marcus=# select jsonb_path_query('{"a": 1}'::JSONB, '$.a'::JSONPATH);
 jsonb_path_query
------------------
 1
(1 row)

Time: 0.875 ms
marcus=# select jsonb_path_query('{"a": 1}'::JSONB, '$.A'::JSONPATH);
 jsonb_path_query
------------------
(0 rows)

Time: 0.629 ms
marcus=# select jsonb_path_query('{"A": 1}'::JSONB, '$.a'::JSONPATH);
 jsonb_path_query
------------------
(0 rows)
```

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/149251#-:-OUAkuF527Ye9n2kXLBi:bmnznme)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @michae2)

___
*[`-- commits` line 6 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/149251#-OUAjY8I2CE1mfKccW4k:-OUAjY8I2CE1mfKccW4l:b23iggr):*
nit: wrap commit messages at 72 characters: https://cockroachlabs.atlassian.net/wiki/spaces/CRDB/pages/73072807/Git+Commit+Messages#How-to-write-a-good-commit-description


<!-- Sent from Reviewable.io -->

ZhouXing19 committed:
jsonpath: allow case insensitivity for keyword idents

Epic: None
Fix: #144255

Previously, if the keyword is not all capital letter, it will cause
syntax error. For example, strIcT $.STRiCt == null will failed the
parsing with at or near "strIcT": syntax error. This is because the
lexer sees strIcT $.STRiCt == null as IDENT $.IDENT == NULL, which is a
rule that is not found in jsonpath.y.

Note that the lexer recognizes strIcT and STRiCt as IDENT by the id, not
by the string. So this commit is to set the ID for all string whose
lowercased form matches a keyword. More specifically, we now do
lval.SetID(lexbase.GetKeywordID(strings.ToLower(lval.Str()))) in the
scanning logic.

So with the same example, strIcT $.STRiCt == null is now recognized as
STRICT $.STRICT == NULL by the parser, for which a rule exists in the
yacc and it will pass the Lexical Analysis. Then, at the Syntactic
Analysis step, the keyword STRICT will be normalized, while the
property accessing part, $.STRiCt, will remain the same, as the
original str STRiCt is still reserved.

Release note (bug fix): Support any casing for
keyword identifiers in a jsonpath query. Note that for 3 special
identifiers "TRUE", "FALSE", "NULL", crdb parses them with case
insensitivity, but PG parses them case sensitively, e.g.
`select '$.active == TrUe'::jsonpath;` will work in crdb but not in
pg.

ZhouXing19 reviewed (commented):
Good point, thanks! Updated the logic to only normalize the keyword idents.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/149251#-:-OUB4tsA9M5RmCOSww4Q:b1kov27)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @mgartner and @michae2)



<!-- Sent from Reviewable.io -->

mgartner commented:
Sorry to throw another wrench in things, but here's another bug:

```sql
-- POSTGRES:
SELECT jsonb_path_query('{"STRICT": 1}'::JSONB, '$.STRICT'::JSONPATH);
--  jsonb_path_query
-- ------------------
--  1
-- (1 row)

-- CRDB with this patch:
SELECT jsonb_path_query('{"STRICT": 1}'::JSONB, '$.STRICT'::JSONPATH);
--   jsonb_path_query
-- --------------------
-- (0 rows)
```

Can the scanner have enough context to know when the ident should be lowered? Might be worth poking around the Postgres source code to see what they do.





<!-- Sent from Reviewable.io -->

yuzefovich reviewed (approved):
Thanks for figuring this out! <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 8 files at r2, 6 of 7 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/149251#-:-OUahWKTB_OmqWMQ7CfJ:b92tadp)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @mgartner and @michae2)

___
*[`pkg/sql/logictest/testdata/logic_test/jsonpath` line 225 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/149251#-OUag3R2003vhILEyB57:-OUag3R2003vhILEyB58:b4r9yq5) ([raw file](https://github.com/cockroachdb/cockroach/blob/bbeb07e0ad12b7cad8b3a942fe58df1c832ea6b8/pkg/sql/logictest/testdata/logic_test/jsonpath#L225)):*
> ```
> $.*
> 
> # keyword "strict" is case-insensitive, but if a key ("STRIcT") will never be normalized
> ```

nit: `s/but if a key/but a key/`.


<!-- Sent from Reviewable.io -->

mgartner reviewed (approved):
Very nice! I'm glad that you found an elegant solution to this!
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 2 of 7 files at r3, 1 of 1 files at r4, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/149251#-:-OUebIsW2uM_9b0rGWQ5:b88uu6a)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @michae2)

___
*[`-- commits` line 31 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/149251#-OUeb8dI9qlHeFdWPL-C:-OUeb8dI9qlHeFdWPL-D:baev1jn):*
Should we open a new issue to track this bug/known limitation?


<!-- Sent from Reviewable.io -->

mgartner commented:
*[`-- commits` line 31 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/149251#-OUeb8dI9qlHeFdWPL-C:-OUeiMSf7-ozmCYnbERC:b-8fx2u7):*
<details><summary><i>Previously, ZhouXing19 (Jane Xing) wrote…</i></summary><blockquote>

Confirmed with Dikshant that we'll keep as this: https://github.com/cockroachdb/cockroach/issues/144255#issuecomment-3019758361
</blockquote></details>

:+1: 


<!-- Sent from Reviewable.io -->
